File: python/782.py
#!/usr/bin/env python3
"""Project Euler 782 - Distinct Rows and Columns.

The complexity of an n x n binary matrix is |{distinct rows} U {distinct columns}|.
c(n,k) = minimum complexity of an n x n binary matrix with exactly k ones.
C(n) = sum_{k=0}^{n^2} c(n,k).

Formula: C(n) = 3*n^2 - 1 - N2 + N4
where N2 = count of k in [1,n^2-1] with c(n,k)=2 (block matrices)
      N4 = count of k in [1,n^2-1] with c(n,k)>=4 (not achievable with comp<=3)

Key constructions for comp<=3:
1. Products: k = d*m for 1 <= d,m <= n-1 (2-type block matrix)
2. Complements: if k is achievable, so is n^2-k (flip all bits)
3. 3x3 kernel matrices: block matrices with 3 row/column groups where
   every column pattern equals some row pattern. These give quadratic
   forms k = sum M[i][j]*n_i*n_j on the simplex n_0+n_1+n_2=n.
"""

import numpy as np
import itertools


def solve(n=10000):
    N = n * n

    # Sieve: achievable[k] = 1 if c(n,k) <= 3
    achievable = np.zeros(N + 1, dtype=np.uint8)
    achievable[0] = 1
    achievable[N] = 1

    # S2: comp=2 values from 2x2 block matrices
    S2 = set()
    for c in range(1, n):
        S2.add(c * c)
        S2.add(N - c * c)
    for x in range(1, n):
        y = n - x
        S2.add(x * x + y * y)
        S2.add(2 * x * y)
    S2.discard(0)
    S2.discard(N)
    N2 = len(S2)
    for k in S2:
        achievable[k] = 1

    # Construction 1: Products d*m with 1 <= d,m <= n-1
    for d in range(1, n):
        achievable[d:d * n:d] = 1

    # Construction 2: Complement symmetry (c(n,k) = c(n,n^2-k))
    achievable_rev = achievable[::-1].copy()
    np.maximum(achievable, achievable_rev, out=achievable)
    del achievable_rev

    # Construction 3: Kernel 3x3 matrices
    # Enumerate all 3x3 binary matrices where every column is also a row
    rows_3 = list(itertools.product([0, 1], repeat=3))
    kernel_forms = set()
    for r0 in rows_3:
        for r1 in rows_3:
            for r2 in rows_3:
                M = [r0, r1, r2]
                cols = [tuple(M[i][j] for i in range(3)) for j in range(3)]
                if all(col in M for col in cols):
                    A = M[0][0]; B = M[1][1]; C = M[2][2]
                    D01 = M[0][1] + M[1][0]
                    D02 = M[0][2] + M[2][0]
                    D12 = M[1][2] + M[2][1]
                    # k(a,b) = aa*a^2 + bb*b^2 + ab*a*b + a1*a + b1*b + c0
                    # where c = n - a - b
                    aa = A + C - D02
                    bb = B + C - D12
                    ab = D01 + 2 * C - D02 - D12
                    a1 = n * (D02 - 2 * C)
                    b1 = n * (D12 - 2 * C)
                    c0 = C * n * n
                    kernel_forms.add((aa, bb, ab, a1, b1, c0))

    for (aa, bb, ab, a1, b1, c0) in kernel_forms:
        for a in range(n + 1):
            b_max = n - a
            if b_max < 0:
                continue
            b = np.arange(b_max + 1, dtype=np.int64)
            k = aa * a * a + bb * b * b + ab * a * b + a1 * a + b1 * b + c0
            valid = (k > 0) & (k < N)
            if valid.any():
                achievable[k[valid]] = 1

    N4 = int(np.sum(achievable[1:N] == 0))
    return 3 * N - 1 - N2 + N4


if __name__ == "__main__":
    print(solve())

File: c/782.c
/*
 * Project Euler 782 - Distinct Rows and Columns
 *
 * C(n) = 3*n^2 - 1 - N2 + N4
 * where N2 = count of k with c(n,k)=2 (block matrices)
 *       N4 = count of k with c(n,k)>=4 (not achievable with comp<=3)
 *
 * Uses bitarray sieve for achievability.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main(void) {
    int n = 10000;
    long long N = (long long)n * n;

    /* Sieve: achievable[k] = 1 if c(n,k) <= 3 */
    unsigned char *achievable = (unsigned char *)calloc(N + 1, 1);
    achievable[0] = 1;
    achievable[N] = 1;

    /* S2: comp=2 values from 2x2 block matrices */
    /* Use a temporary bitset for S2 */
    unsigned char *is_s2 = (unsigned char *)calloc(N + 1, 1);

    for (int c = 1; c < n; c++) {
        long long v = (long long)c * c;
        if (v > 0 && v < N) is_s2[v] = 1;
        long long w = N - v;
        if (w > 0 && w < N) is_s2[w] = 1;
    }
    for (int x = 1; x < n; x++) {
        int y = n - x;
        long long v1 = (long long)x * x + (long long)y * y;
        long long v2 = 2LL * x * y;
        if (v1 > 0 && v1 < N) is_s2[v1] = 1;
        if (v2 > 0 && v2 < N) is_s2[v2] = 1;
    }

    int N2 = 0;
    for (long long k = 1; k < N; k++) {
        if (is_s2[k]) {
            N2++;
            achievable[k] = 1;
        }
    }
    free(is_s2);

    /* Construction 1: Products d*m with 1 <= d,m <= n-1 */
    for (int d = 1; d < n; d++) {
        for (long long k = d; k < (long long)d * n; k += d) {
            achievable[k] = 1;
        }
    }

    /* Construction 2: Complement symmetry */
    for (long long k = 1; k < N; k++) {
        if (achievable[k] || achievable[N - k]) {
            achievable[k] = 1;
            achievable[N - k] = 1;
        }
    }

    /* Construction 3: Kernel 3x3 matrices
     * Enumerate all 3x3 binary matrices where every column is also a row
     */
    int rows_3[8][3];
    for (int i = 0; i < 8; i++) {
        rows_3[i][0] = (i >> 2) & 1;
        rows_3[i][1] = (i >> 1) & 1;
        rows_3[i][2] = i & 1;
    }

    /* Store unique kernel forms (aa, bb, ab, a1, b1, c0) */
    typedef struct { int aa, bb, ab, a1, b1; long long c0; } KForm;
    KForm forms[600];
    int nforms = 0;

    for (int r0i = 0; r0i < 8; r0i++) {
        for (int r1i = 0; r1i < 8; r1i++) {
            for (int r2i = 0; r2i < 8; r2i++) {
                int M[3][3];
                for (int j = 0; j < 3; j++) {
                    M[0][j] = rows_3[r0i][j];
                    M[1][j] = rows_3[r1i][j];
                    M[2][j] = rows_3[r2i][j];
                }

                /* Check that every column is also a row */
                int ok = 1;
                for (int j = 0; j < 3 && ok; j++) {
                    int col[3] = { M[0][j], M[1][j], M[2][j] };
                    int found = 0;
                    for (int ri = 0; ri < 3 && !found; ri++) {
                        if (M[ri][0] == col[0] && M[ri][1] == col[1] && M[ri][2] == col[2])
                            found = 1;
                    }
                    if (!found) ok = 0;
                }
                if (!ok) continue;

                int A = M[0][0], B = M[1][1], C = M[2][2];
                int D01 = M[0][1] + M[1][0];
                int D02 = M[0][2] + M[2][0];
                int D12 = M[1][2] + M[2][1];
                int aa = A + C - D02;
                int bb = B + C - D12;
                int ab = D01 + 2 * C - D02 - D12;
                int a1 = n * (D02 - 2 * C);
                int b1 = n * (D12 - 2 * C);
                long long c0 = (long long)C * n * n;

                /* Check if this form is already recorded */
                int dup = 0;
                for (int fi = 0; fi < nforms && !dup; fi++) {
                    if (forms[fi].aa == aa && forms[fi].bb == bb &&
                        forms[fi].ab == ab && forms[fi].a1 == a1 &&
                        forms[fi].b1 == b1 && forms[fi].c0 == c0)
                        dup = 1;
                }
                if (!dup && nforms < 600) {
                    forms[nforms].aa = aa;
                    forms[nforms].bb = bb;
                    forms[nforms].ab = ab;
                    forms[nforms].a1 = a1;
                    forms[nforms].b1 = b1;
                    forms[nforms].c0 = c0;
                    nforms++;
                }
            }
        }
    }

    for (int fi = 0; fi < nforms; fi++) {
        int aa = forms[fi].aa, bb = forms[fi].bb, ab = forms[fi].ab;
        int a1 = forms[fi].a1, b1 = forms[fi].b1;
        long long c0 = forms[fi].c0;
        for (int a = 0; a <= n; a++) {
            int b_max = n - a;
            for (int b = 0; b <= b_max; b++) {
                long long k = (long long)aa * a * a + (long long)bb * b * b +
                              (long long)ab * a * b + (long long)a1 * a +
                              (long long)b1 * b + c0;
                if (k > 0 && k < N)
                    achievable[k] = 1;
            }
        }
    }

    int N4 = 0;
    for (long long k = 1; k < N; k++) {
        if (achievable[k] == 0) N4++;
    }

    printf("%lld\n", 3 * N - 1 - N2 + N4);

    free(achievable);
    return 0;
}
